src/utils/data-access.ts

```diff
@@ -0,0 +1,98 @@
+import { PromiseResult } from 'aws-sdk/lib/request';
+import { AWSError } from 'aws-sdk/lib/error';
+import DynamoDB, { DocumentClient, QueryOutput } from 'aws-sdk/clients/dynamodb';
+import { Configurations } from './configuration';
+import * as domain from '../domain';
+
+export class DataAccess {
+  private static instance: DataAccess;
+
+  private readonly tableName: string = Configurations.getInstance().dynamoTableName;
+
+  private static docClient: DocumentClient;
+
+  private constructor() {
+    console.log('config', Configurations.getInstance().dynamoTableName);
+    DataAccess.docClient = new DynamoDB.DocumentClient(Configurations.getInstance().dynamoParams);
+  }
+
+  public static getInstance(): DataAccess {
+    if (!DataAccess.instance) {
+      DataAccess.instance = new DataAccess();
+    }
+
+    return DataAccess.instance;
+  }
+
+  public getById(options: {
+    indexName: string;
+    keyCondition: string;
+    expressionAttributeValues: { [key: string]: string };
+  }): Promise<PromiseResult<QueryOutput, AWSError>> {
+    const params = {
+      TableName: this.tableName,
+      IndexName: options.indexName,
+      KeyConditionExpression: options.keyCondition,
+      ExpressionAttributeValues: options.expressionAttributeValues,
+    };
+    console.log('params getbyId', params);
+    return DataAccess.docClient.query(params).promise();
+  }
+
+  public async put<T>(
+    payload: T,
+    options?: { conditionalExpression: string; expressionAttributeValues: { [key: string]: string } },
+  ): Promise<PromiseResult<DocumentClient.PutItemOutput, AWSError>> {
+    let query = {
+      TableName: this.tableName,
+      Item: payload,
+      ConditionExpression: undefined,
+      ExpressionAttributeValues: undefined,
+    };
+    if (options) {
+      query = {
+        ...query,
+        ConditionExpression: options.conditionalExpression,
+        ExpressionAttributeValues: options.expressionAttributeValues,
+      };
+    }
+    console.log(DataAccess.docClient);
+    return DataAccess.docClient.put(query).promise();
+  }
+
+  public async upsertTrailerRegistration(
+    trailerRegistration: domain.TrailerRegistration,
+  ): Promise<PromiseResult<DocumentClient.PutItemOutput, AWSError>> {
+    return DataAccess.getInstance().put<domain.TrailerRegistration>(trailerRegistration);
```

Comment:
does ```this.``` substitute for ```DataAccess.getInstance()``` Seems they both refer to the same instance.  There are a few occurances where the code could be simplified.

Reply:
replaced all instances.